Add mcumgr shell management

bpbradley opened the pull request:
I just submitted a PR to the upstream mcumgr repo (https://github.com/apache/mynewt-mcumgr/pull/99) which implements the shell command handler for mcumgr.

I am preparing this PR in hopes that it gets merged, and also to get feedback on the implementation. Included in this PR is a change to the mcumgr smp_srv sample which includes an additional overlay allowing this handler to be enabled.

**To test:**

1. `west update` (to update the mcumgr lib to the one pointed to by my PR in the upstream repo)
1. Build and flash mcuboot
1. Build, sign, and flash smp_srv sample with `OVERLAY_CONFIG=overlay-shell-mgmt.conf`. ~~**Note:** Don't use the shell as a backend, as this won't work without changing the shell-mgmt overlay. Details on this below.~~
1. Test shell management feature with mcumgr-cli.

An example test with mcumgr-cli would be: `mcumgr -c <connection profile> shell exec "kernel cycles"`

To be done:

- [x] Update west.yml to point to the official commit if / when the PR is merged

- [x] Update docs to show that this command handler is supported?

Edit: I forgot to mention that in order to use this you need to perform a `west update` so that the intended mcumgr source is used.

Edit 2: After some feedback I have decided to remove the support for serial backend for this feature to prevent sending anything to the output of another interface. The dummy backend is now enforced.

Edit 3: I realized that with the dummy backend enforced and fully isolated from any physical transport layer for the shell, there is no reason to require that shell transport is disabled. It is possible now to have this work safely in coordination with a serial shell, and it will even work with shell as the transport layer for mcumgr.

bpbradley committed:
mgmt: mcumgr: add configuration option to enable shell management

This configuration is be required to enable the shell management
command handler supported by the core mcumgr library

Signed-off-by: Brian Bradley <brian.bradley.p@gmail.com>

de-nordic commented:
OK since you are doing more changes than just the west, give west separate commit.
Change the title of the review accordingly: you are not only updating mcumgr, you are also enabling it in sample.
Thanks.

bpbradley commented:
> OK since you are doing more changes than just the west, give west separate commit.
> Change the title of the review accordingly: you are not only updating mcumgr, you are also enabling it in sample.
> Thanks.

Ah, sorry about that. I wasn't sure if I should include the sample but figured it would be helpful for testing. I will update the commits accordingly. Where should the Kconfig change go? It is required to enable the feature in mcumgr. Also a separate commit?

bpbradley committed:
samples: add support for shell management in mcumgr sample

The shell management feature must be enabled using the provided overlay

Signed-off-by: Brian Bradley <brian.bradley.p@gmail.com>

bpbradley committed:
west: update revision for mcumgr library

The updated revision includes the ability to use shell management

Signed-off-by: Brian Bradley <brian.bradley.p@gmail.com>

de-nordic commented:
> Ah, sorry about that. I wasn't sure if I should include the sample but figured it would be helpful for testing.

It is nice that you have updated the sample because it actually makes use of the PR.

> Where should the Kconfig change go? It is required to enable the feature in mcumgr. Also a separate commit?

You may create additional commit just for the Kconfig and then sample on top.

> Also I assume I should convert this PR to a draft? And what title prefix would be appropriate in this case? modules: mcumgr perhaps?

I think that there is no need for draft. You are not modifying mcumgr  module here only west, Kconfig and sample, figure out prefixes with git log. Reviewers will let you know if something will be off.

bpbradley commented:
> > Ah, sorry about that. I wasn't sure if I should include the sample but figured it would be helpful for testing.
> 
> It is nice that you have updated the sample because it actually makes use of the PR.
> 
> > Where should the Kconfig change go? It is required to enable the feature in mcumgr. Also a separate commit?
> 
> You may create additional commit just for the Kconfig and then sample on top.
> 
> > Also I assume I should convert this PR to a draft? And what title prefix would be appropriate in this case? modules: mcumgr perhaps?
> 
> I think that there is no need for draft. You are not modifying mcumgr module here only west, Kconfig and sample, figure out prefixes with git log. Reviewers will let you know if something will be off.

Thanks for the input, I believe I have my commits done correctly now, and I just removed the prefix from the PR title since I'm not sure if there is a single area which is appropriate for the full PR. I suppose now we just wait to see if the PR in the apache repo is merged or if there is any feedback on the implementation. Thanks for the help!

Also am I able to add the DNM label myself? Or is that only something that members and collaborators can do?

de-nordic commented:
> 
> Also am I able to add the DNM label myself? Or is that only something that members and collaborators can do?

Yes you may add DNM by yourself. You may add any labels that you find relevant.

de-nordic commented:
Could you give some screenshot from console that would show how you use the command, on the smp_svr sample, and how it responds? As a demonstration. Thanks.

bpbradley commented:
> Could you give some screenshot from console that would show how you use the command, on the smp_svr sample, and how it responds? As a demonstration. Thanks.

Sure, the basic idea is that you can execute any shell command as if you had the shell open. So any registered shell commands can be used without actually having physical access to the shell backend (for instance, internal serial which is inaccessible). This could be useful to control/manage devices through a single management interface (mcumgr). It also provides a means of accessing the shell over bluetooth since that is a supported transport for mcumgr, but not a supported backend for the shell.

Here is a screenshot of using this on a basic kernel shell command.

![image](https://user-images.githubusercontent.com/64746189/97030545-e9315200-152c-11eb-9868-8c5188069303.png)

nvlsianpu commented on subsys/mgmt/mcumgr/Kconfig:
```diff
@@ -123,6 +123,12 @@ config FS_MGMT_PATH_SIZE
 
+config MCUMGR_CMD_SHELL_MGMT
+	bool "Enable mcumgr handlers for shell management"
+	select SHELL
```
```suggestion
	depends on !SHELL_BACKEND_SERIA && SHELL
	select SHELL_BACKEND_DUMMY
```
^^ This makes impossible to misconfigure the feature you added.
You can also add same comments as in overlay-shell-mgmt.conf

bpbradley replied:
Thanks for the feedback!

So the way I have designed it, you are allowed to use `SHELL_BACKEND_SERIAL` and I check for it in the mcumgr command handler. The only "issue" is that you can't get the shell output when using the serial backend. The command is still executed, and you still get the return code, and the shell output is still sent through the serial output, but mcumgr can't get it. So basically I don't enforce the dummy backend, it is just a more complete implementation if you are using it.

If that is confusing, I am okay with just enforcing the dummy backend and I would change the implementation in the mcumgr PR.

However you bring up a good point because it is going to be impossible to use `CONFIG_MCUMGR_SMP_SHELL` with this, so I should have `depends on !CONFIG_MCUMGR_SMP_SHELL` for sure.

nvlsianpu replied:
So `imply SHELL_BACKEND_DUMMY` might be used instead (see https://docs.zephyrproject.org/latest/guides/kconfig/tips.html?highlight=imply#the-imply-statement).

nvlsianpu replied:
@de-nordic Make it sense to you that the shell response arrives by another channel? I afraid that this might be notices as a bug by the user.

de-nordic replied:
While returning result code may be useful even if no command output gets received, e.g. while toggling some parameters, reporting output to other channel may break something on user side.

bpbradley replied:
That makes sense. So probably the best solution is to just enforce the dummy backend. I will change this in the implementation and then make the changes suggested here. Thanks!

de-nordic replied:
The mcumgr stat command seems to be able to pack, as cbor, output on device and read it on host via mcumgr. Wouldn't it be possible to do the same here?

bpbradley replied:
@de-nordic @nvlsianpu  Thanks again for the feedback. I went ahead and removed the support for serial backend from mcumgr, so as to not introduce confusing behavior where it isn't needed. I then changed the implementation in this repo now to reflect that and used the suggestions provided here to prevent misconfiguration. 

Dummy backend is now enforced.

nvlsianpu reviewed (commented):
Nice! This is outstanding feature.

de-nordic commented on subsys/mgmt/mcumgr/Kconfig:
```diff
@@ -123,6 +123,14 @@ config FS_MGMT_PATH_SIZE
 	  and download commands.
 endif
 
+# Must use the dummy backend for the shell so that mcumgr can capture output
```
Move that comment into help for the option.

bpbradley replied:
That makes sense. Will do. Thanks!

de-nordic reviewed (changes requested):
Minor issues.

I have been able to  successfully test this feature on nrf52840dk.

bpbradley commented on subsys/mgmt/mcumgr/Kconfig:
```diff
@@ -123,6 +123,14 @@ config FS_MGMT_PATH_SIZE
+	depends on !SHELL_BACKEND_SERIAL && SHELL
+	select SHELL_BACKEND_DUMMY
+	help
+	  Enables mcumgr handlers for shell management
```
That is a good idea. I will update this later today.

bpbradley replied:
I am actually thinking about this more, and I realize that the dummy and serial shell backends don't really have a problem co-existing from what I can see, so long as I am explicit in getting the correct backend pointer when accessing the shell from within mcumgr. 

This means that the serial shell can exist, it can even be used as the backend for mcumgr transport I think (however this would be a bit of a strange use case), and both backends can be used independently without unexpected interactions. 

I think it might make sense to just use the dummy output, and allow the serial one to exist (given that I don't touch the serial backend within mcumgr). I will do some tests on this to be sure I am not missing something.

bpbradley replied:
@de-nordic  sorry for the back and forth, but I think this new implementation which doesn't dictate the shell configuration works better. Serial backend is again allowed, but it won't interact with (is independent of) mcumgr which still uses the dummy backend.

de-nordic replied:
> Serial backend is again allowed, but it won't interact with (is independent of) mcumgr which still uses the dummy backend.

Right, I agree, I have tested the feature again with UART shell and Bluetooth. Everything works fine and there is no interference.

de-nordic commented on subsys/mgmt/mcumgr/Kconfig:
```diff
@@ -123,6 +123,15 @@ config FS_MGMT_PATH_SIZE
+	depends on SHELL
+	select SHELL_BACKEND_DUMMY
+	help
+	  Enables mcumgr handlers for shell management. Must use the dummy backend
```
`Must use the dummy backend ...` -> `The handler utilizes the dummy shell back-end to capture (...)`

or something like that, reword it little bit.

bpbradley replied:
Makes sense. Will do. Thanks!

bpbradley replied:
I believe this is all taken care of now. I just updated the language according to your recommendations, and added a bit more clarifying detail.

de-nordic reviewed (approved):
I left some nitpicking comments regarding the Help in the Kconfig for the `MCUMGR_CMD_SHELL_MGMT`.

Anyway, I have tested the feature on nrf52840dk_nrf52840 with UART shell attached and Bluetooth enabled; I have been able to successfully execute shell commands, via Bluetooth, with mcumgr and receive command output and result codes. No interference with UART shell.

Generally approving while waiting for the mynewt for PR to go through.

utzig commented on west.yml:
```diff
@@ -95,8 +97,9 @@ manifest:
-      revision: 1e5e3ecd5b836183e9970991ab2c8f38c5e60298
+    - name: mynewt-mcumgr
+      remote: mynewt
+      revision: pull/99/head
```
Mynewt remote can be dropped, and update revision to `pull/31/head` (https://github.com/zephyrproject-rtos/mcumgr/pull/31)

bpbradley replied:
Will do. Should I update the docs here: https://docs.zephyrproject.org/latest/guides/device_mgmt/ to indicate that shell management is supported as well? Or leave that for a future PR?

nvlsianpu replied:
@de-nordic ^^ Where to document this?

de-nordic replied:
> 
> 
> Will do. Should I update the docs here: https://docs.zephyrproject.org/latest/guides/device_mgmt/ to indicate that shell management is supported as well? Or leave that for a future PR?

Yes, please update the documentation.

de-nordic replied:
@bpbradley The mcumgr module changes have been merged to zephyrproject-rtos/mcumgr, withi commit https://github.com/zephyrproject-rtos/mcumgr/commit/f28a637db12c4b12fb2b18bddc6b2a0deaa95251

bpbradley replied:
Just pushed the changes reflecting the merged commit, and I have updated the docs in a separate commit. This should be ready for merge (I cant seem to be able to remove the DNM label).

bpbradley committed:
doc: document support of shell management for mcumgr

This is documented similarly to how other command
handlers are documented. This documents that shell
management is supported, but relies on the example
project for implementation details.

Signed-off-by: Brian Bradley <brian.bradley.p@gmail.com>

aunsbjerg reviewed (approved):
Nice feature! Changes LGTM, only thing I would request is to add a bit of info in the sample documentation on how to build and use the shell mgmt bits.

bpbradley commented:
> Nice feature! Changes LGTM, only thing I would request is to add a bit of info in the sample documentation on how to build and use the shell mgmt bits.

Ah good point. I can add that later today. Thanks!
